### ISSUE TITLE
Included assets are not deserialized

### ISSUE THREAD

**Issue by aKzenT** [contentful/contentful.net#39]
When querying entries which link to assets, my expectation would be that the assets will be deserialized automatically, just as it works with assets. However it seems that the automatic link resolution only works for entries and not for assets.

I believe the error is in the following lines in the GetEntriesAsync method:

```
            var includedLinks = json.SelectTokens("$.items..fields..sys").ToList();
            includedLinks.AddRange(json.SelectTokens("$.includes.Entry..fields..sys"));
```

As you can see, only entry-includes are considered, not asset-includes.

**Comment by Roblinde** [comment]
Your expectation is correct, this should work for assets as well. Could you share your content type so I can have a quick look? It seems to be working correctly for me.

The reason that assets do not need to be included in the code you mention above is that they already implement a specific json-converter that should take care of this automatically.

**Comment by aKzenT** [comment]
The ContentType has various levels of nesting, but it basically breaks down to something like:

```
class Container
{
     SubContainer[] Items;
}

class SubContainer
{
    Item[] Items;
}

class Item
{
    Asset Media;
}
```

I understand that there is a AssetJsonConverter, however if I understand the code correctly, this assumes that the original Json from Contentful was modified so that the links to included assets (and entries) are replaced by the actual content. This seems to happen for entries, but not for assets.

**Comment by Roblinde** [comment]
Ah, so the Asset you're trying to get is referenced from `Container.SubContainer[0].Items[0].Media`

I assume you have the correct level of includes set on your `QueryBuilder`?

I will take a look right away, you might be right that there's something fishy here...

**Comment by aKzenT** [comment]
Yes. I set include level extra high and made sure that the asset is included in the returned JSON.

**Comment by Roblinde** [comment]
Having a look now, get back to you shortly.

**Comment by Roblinde** [comment]
OK, so I've had a look and found a few flaws in the approach that I have in the SDK right now.

Basically you're on the right track about the 

```
var includedLinks = json.SelectTokens("$.items..fields..sys").ToList();
includedLinks.AddRange(json.SelectTokens("$.includes.Entry..fields..sys"));
```

being the issue, but not quite the way you thought.

The lines are correct, there should be no assets in there. These lines are to figure out which tokens to replace and only entries can have references to other content, hence only entries are being collected here.

However, since the Contentful API basically returns it's content in three separate collections. `items`, `Includes.Entry` and `Includes.Asset` what we're trying to do here is to replace tokens in items with tokens from the `Includes` section to make them easily serializable. But the problem is that the order of moving these tokens matter. What happens in your case (and my test case I've setup) is that a token from the `Includes.Entry` section is moved up to the `items` section, but it has not yet had it's own references resolved.

I'll be looking into a more robust way of solving this. If you need a workaround in the meantime another solution could be using the `GetEntriesCollection` method which includes two separate collections for `IncludedEntries` and `IncludedAssets`.

**Comment by aKzenT** [comment]
Okay, thank you for your rapid feedback.

Just as an idea... JSON.NET also supports references to allow circular structures etc. if you change the JSON to include $id and $ref tokens, maybe that could be an easier way to support includes instead of having to move the tokens arround.

Or, if you want to stick with the current model, wouldn't it be sufficient to only start with links in the items collection and then recursiveley replace tokens in the replacement tokens themselves?

**Comment by Roblinde** [comment]
hmm, $id and $ref tokens might not be a bad idea at all. Could save me some headache when deserializing. I'll look into that.

The more I look into this the more it actually seems to be working as intended. The bug I stumbled upon was if you reference objects of the same type in the object graph. Are you doing that in your example?

**Comment by aKzenT** [comment]
Hm, I'm not sure I understand what you mean by that, but in my example there can be multiple containers and multiple items, all having an asset. So I guess, yes, I'm referencing objects of the same type...

**Comment by Roblinde** [comment]
Yeah, sorry I was being unclear. I created a class `Thing` which had a reference of `List<Thing>` so both of the same type. This caused a problem when deserializing.

Is it just the Asset property that is null for you or other fields as well?

**Comment by aKzenT** [comment]
I only noticed the Asset property so far.

**Comment by Roblinde** [comment]
I've been investigating $ref and $id a bit further and I'd love to go down that path. There are a number of challenges though.

Firstly, the $id must come before the $ref in the Json structure or it is not recognized. Furthermore, you can't deserialize a fragment of the json as I normally do, because it needs to be able to find the referenced object.

That said though I think that approach would be much cleaner and I'll do some further testing and see how I could implement that. It won't be a quick fix I'm afraid though. 

For your issue I think I need some more info to dig into. Could you perhaps paste a gist with some sample json response? Just make sure to remove anything sensitive.

**Comment by aKzenT** [comment]
Hey,

I sent you a pull request with a failing test case.

The json is a redacted version of my problem and I see two issues here, both present in my actual project and in the test case.

First the Asset is lost as described above (only SystemProperties present). Second I would only expect a single result from the function, as there is only one entry (+includes) in the JSON. But it seems there are also several other items returned. I first thought this was by design, but thinking about this, I believe it might be a bug as well.

**Comment by Roblinde** [comment]
Perfect, many thanks. I'll have a look at it first thing tomorrow.

**Comment by aKzenT** [comment]
Hm, I think the second issue is caused by my models having an "items" property, which is picked up in some selectors where you use "$..items" instead of "$.items".

**Comment by Roblinde** [comment]
Yes noticed that. Just as you suggested I removed the superflous ".". I actually managed to get the test to pass already by making sure assets are always handled before entries in the replacement algorithm. I need to ponder a bit more about potential implications of this though... I'll dive deeper tomorrow.

**Comment by aKzenT** [comment]
If you want to give the $id/$ref a try I created a PoC in #43 

I believe that all other (JSON based) solutions will fail in some scenario or another, since reference loops cannot be expressed in another way in JSON.

**Comment by Roblinde** [comment]
This should now be resolved in PR #44 🎉 

The issue will be left open until an updated package has been released.
